### PR TITLE
CHB optimization

### PR DESCRIPTION
**Pull request by noescape00**
Slightly optimized CHB. This is the 1st PR to get CHB ready for the CHT and CM activation. 

Tests pass, networks synced

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreBehavior.cs [hunk 1]
```diff
@@ -176,6 +179,23 @@ protected override void DetachCore()
             this.logger.LogTrace("(-)");
         }
 
+        /// <summary>Gets the highest header between the header of the last block that was sent to a peer and the expected peer's tip.</summary>
+        private ChainedHeader GetHighestHeader(INetworkPeer peer)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(peer), peer.RemoteSocketEndpoint);
+
+            var chainBehavior = peer.Behavior<ChainHeadersBehavior>();
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.BlockStore/BlockStoreBehavior.cs]: I tend to set `chainBehavior` as a field on the behaviour and set it on `AttachCore`

File: src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs [hunk 2]
```diff
@@ -33,62 +30,23 @@ public class ChainHeadersBehavior : NetworkPeerBehavior
...
-                return this.Chain.GetBlock(tip.HashBlock) ?? tip;
-            }
-        }
-
-        /// <summary><c>true</c> to respond to "getheaders" messages, <c>false</c> to ignore it.</summary>
-        public bool CanRespondToGetHeaders { get; set; }
+        /// <remarks>
+        /// The announced tip is accepted if it seems to be valid. Validation is only done on headers and so the announced tip may refer to invalid block.
+        /// </remarks>
+        public ChainedHeader ExpectedTip { get; private set; }
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs]: Should this not be `PeerTip`?

Comment (noescape00) [reply src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs]: That's expected peer tip. 
We were arguing with apro on what name suits best, ended up with ExpectedTip. 
I can rename to ExpectedPeerTip if you'd like.

File: src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs [hunk 3]
```diff
@@ -373,47 +313,29 @@ protected override void DetachCore()
...
 
             // If we made any advancement or the sync is enforced by 'doTrySync'- continue syncing.
-            if (doTrySync || (this.pendingTip == null) || (pendingTipBefore == null) || (pendingTipBefore.HashBlock != this.pendingTip.HashBlock))
+            if (doTrySync || (this.ExpectedTip == null) || (pendingTipBefore == null) || (pendingTipBefore.HashBlock != this.ExpectedTip.HashBlock))
                 await this.TrySyncAsync().ConfigureAwait(false);
 
             this.logger.LogTrace("(-)");
         }
 
-        public void SetPendingTip(ChainedHeader newTip)
```
Comment (dangershony) [inline src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs]: Is this method not used?

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreBehavior.cs (hunk 1 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.BlockStore/BlockStoreBehavior.cs]: Yeah. This whole behavior requires a refactoring, it's a big mess. 

Will fix this one.

File: src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs (hunk 2 above)
Comment (dangershony) [reply src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs]: Ok if there was an argument then I will just stay out of it leave as is.

Comment (noescape00) [reply src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs]: Not really an argument, just didn't know what would be the name. 

ExpectedPeerTip sounds good too, will fix.

File: src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs (hunk 3 above)
Comment (noescape00) [reply src/Stratis.Bitcoin/Base/ChainHeadersBehavior.cs]: No. It used to be when BS behavior was using pending tip. Now we have 2 tips (one represents the last block that was sent to the peer and another represents last header received from the peer). 
BTC Core do that so we decided to copy that logic.

**Comment by noescape00** [comment]
@dangershony applied fixes.

**Comment by noescape00** [comment]
Synced the networks 2nd time just in case.

**Pull request merged**